File: advent_of_code/pkg/aoc/grid.go
package aoc

import "strings"

// Point represents a 2D coordinate
type Point struct {
	X, Y int
}

func (p Point) Add(other Point) Point {
	return Point{p.X + other.X, p.Y + other.Y}
}

func (p Point) Sub(other Point) Point {
	return Point{p.X - other.X, p.Y - other.Y}
}

// Manhattan returns the Manhattan distance from origin
func (p Point) Manhattan() int {
	return Abs(p.X) + Abs(p.Y)
}

// ManhattanTo returns the Manhattan distance to another point
func (p Point) ManhattanTo(other Point) int {
	return Abs(p.X-other.X) + Abs(p.Y-other.Y)
}

// Neighbors4 returns the 4 cardinal neighbors (up, down, left, right)
func (p Point) Neighbors4() []Point {
	return []Point{
		{p.X, p.Y - 1}, // up
		{p.X, p.Y + 1}, // down
		{p.X - 1, p.Y}, // left
		{p.X + 1, p.Y}, // right
	}
}

// Neighbors8 returns all 8 neighbors including diagonals
func (p Point) Neighbors8() []Point {
	return []Point{
		{p.X, p.Y - 1},     // up
		{p.X, p.Y + 1},     // down
		{p.X - 1, p.Y},     // left
		{p.X + 1, p.Y},     // right
		{p.X - 1, p.Y - 1}, // up-left
		{p.X + 1, p.Y - 1}, // up-right
		{p.X - 1, p.Y + 1}, // down-left
		{p.X + 1, p.Y + 1}, // down-right
	}
}

// Direction constants
var (
	Up    = Point{0, -1}
	Down  = Point{0, 1}
	Left  = Point{-1, 0}
	Right = Point{1, 0}

	// Cardinal directions
	Directions4 = []Point{Up, Down, Left, Right}

	// All 8 directions
	Directions8 = []Point{
		Up, Down, Left, Right,
		{-1, -1},
		{1, -1},
		{-1, 1},
		{1, 1},
	}
)

// TurnLeft rotates a direction 90 degrees counter-clockwise
func TurnLeft(dir Point) Point {
	return Point{dir.Y, -dir.X}
}

// TurnRight rotates a direction 90 degrees clockwise
func TurnRight(dir Point) Point {
	return Point{-dir.Y, dir.X}
}

// Grid represents a 2D grid of runes
type Grid struct {
	Data   [][]rune
	Width  int
	Height int
}

func NewGrid(lines []string) *Grid {
	if len(lines) == 0 {
		return &Grid{}
	}

	height := len(lines)
	width := len(lines[0])
	data := make([][]rune, height)

	for y, line := range lines {
		data[y] = []rune(line)
		if len(data[y]) > width {
			width = len(data[y])
		}
	}

	return &Grid{
		Data:   data,
		Width:  width,
		Height: height,
	}
}

func NewGridFromString(s string) *Grid {
	return NewGrid(Lines(s))
}

// Get returns the rune at position, or 0 (null rune) if out of bounds.
func (g *Grid) Get(p Point) rune {
	if !g.InBounds(p) {
		return 0
	}
	if p.X >= len(g.Data[p.Y]) {
		return 0
	}
	return g.Data[p.Y][p.X]
}

func (g *Grid) Set(p Point, r rune) {
	if g.InBounds(p) {
		g.Data[p.Y][p.X] = r
	}
}

func (g *Grid) InBounds(p Point) bool {
	return p.X >= 0 && p.X < g.Width && p.Y >= 0 && p.Y < g.Height
}

// Find returns the first position of the given rune, or (-1,-1) if not found
func (g *Grid) Find(r rune) Point {
	for y, row := range g.Data {
		for x, cell := range row {
			if cell == r {
				return Point{x, y}
			}
		}
	}
	return Point{-1, -1}
}

// FindAll returns all positions containing the given rune
func (g *Grid) FindAll(r rune) []Point {
	var result []Point
	for y, row := range g.Data {
		for x, cell := range row {
			if cell == r {
				result = append(result, Point{x, y})
			}
		}
	}
	return result
}

func (g *Grid) ForEach(fn func(p Point, r rune)) {
	for y, row := range g.Data {
		for x, cell := range row {
			fn(Point{x, y}, cell)
		}
	}
}

// Copy creates a deep copy of the grid
func (g *Grid) Copy() *Grid {
	data := make([][]rune, g.Height)
	for y, row := range g.Data {
		data[y] = make([]rune, len(row))
		copy(data[y], row)
	}
	return &Grid{
		Data:   data,
		Width:  g.Width,
		Height: g.Height,
	}
}

func (g *Grid) String() string {
	var sb strings.Builder
	for _, row := range g.Data {
		sb.WriteString(string(row))
		sb.WriteRune('\n')
	}
	return sb.String()
}

// Count returns the number of cells matching the predicate
func (g *Grid) Count(fn func(r rune) bool) int {
	count := 0
	g.ForEach(func(p Point, r rune) {
		if fn(r) {
			count++
		}
	})
	return count
}

func (g *Grid) CountRune(r rune) int {
	return g.Count(func(c rune) bool { return c == r })
}

File: projects/advent_of_code/python/aoc/grid.py
"""Grid and 2D coordinate utilities for Advent of Code."""

from __future__ import annotations
from dataclasses import dataclass
from typing import Callable, Iterator

__all__ = [
    "Point",
    "Grid",
    "UP",
    "DOWN",
    "LEFT",
    "RIGHT",
    "DIRECTIONS4",
    "DIRECTIONS8",
    "turn_left",
    "turn_right",
]


@dataclass(frozen=True, slots=True)
class Point:
    """A 2D coordinate."""

    x: int
    y: int

    def __add__(self, other: Point) -> Point:
        return Point(self.x + other.x, self.y + other.y)

    def __sub__(self, other: Point) -> Point:
        return Point(self.x - other.x, self.y - other.y)

    def __mul__(self, scalar: int) -> Point:
        return Point(self.x * scalar, self.y * scalar)

    def __neg__(self) -> Point:
        return Point(-self.x, -self.y)

    def manhattan(self) -> int:
        """Manhattan distance from origin."""
        return abs(self.x) + abs(self.y)

    def manhattan_to(self, other: Point) -> int:
        """Manhattan distance to another point."""
        return abs(self.x - other.x) + abs(self.y - other.y)

    def neighbors4(self) -> list[Point]:
        """Return 4 cardinal neighbors (up, down, left, right)."""
        return [
            Point(self.x, self.y - 1),  # up
            Point(self.x, self.y + 1),  # down
            Point(self.x - 1, self.y),  # left
            Point(self.x + 1, self.y),  # right
        ]

    def neighbors8(self) -> list[Point]:
        """Return all 8 neighbors including diagonals."""
        return [
            Point(self.x, self.y - 1),  # up
            Point(self.x, self.y + 1),  # down
            Point(self.x - 1, self.y),  # left
            Point(self.x + 1, self.y),  # right
            Point(self.x - 1, self.y - 1),  # up-left
            Point(self.x + 1, self.y - 1),  # up-right
            Point(self.x - 1, self.y + 1),  # down-left
            Point(self.x + 1, self.y + 1),  # down-right
        ]

    def __iter__(self):
        """Allow unpacking: x, y = point"""
        yield self.x
        yield self.y


# Direction constants
UP = Point(0, -1)
DOWN = Point(0, 1)
LEFT = Point(-1, 0)
RIGHT = Point(1, 0)

DIRECTIONS4 = [UP, DOWN, LEFT, RIGHT]
DIRECTIONS8 = [
    UP,
    DOWN,
    LEFT,
    RIGHT,
    Point(-1, -1),
    Point(1, -1),
    Point(-1, 1),
    Point(1, 1),
]


def turn_left(direction: Point) -> Point:
    """Rotate direction 90 degrees counter-clockwise."""
    return Point(direction.y, -direction.x)


def turn_right(direction: Point) -> Point:
    """Rotate direction 90 degrees clockwise."""
    return Point(-direction.y, direction.x)


class Grid:
    """A 2D grid of characters."""

    def __init__(self, data: list[list[str]]):
        self.data = data
        self.height = len(data)
        self.width = max(len(row) for row in data) if data else 0

    @classmethod
    def from_string(cls, s: str) -> Grid:
        """Create grid from multi-line string."""
        lines = s.rstrip("\n").split("\n")
        return cls([list(line) for line in lines])

    @classmethod
    def from_lines(cls, lines: list[str]) -> Grid:
        """Create grid from list of strings."""
        return cls([list(line) for line in lines])

    @classmethod
    def empty(cls, width: int, height: int, fill: str = ".") -> Grid:
        """Create an empty grid filled with a character."""
        return cls([[fill] * width for _ in range(height)])

    def get(self, p: Point, default: str = "") -> str:
        """Get character at position, or default if out of bounds."""
        if not self.in_bounds(p):
            return default
        if p.x >= len(self.data[p.y]):
            return default
        return self.data[p.y][p.x]

    def __getitem__(self, p: Point) -> str:
        """Get character at position. Raises IndexError if out of bounds."""
        if not self.in_bounds(p):
            raise IndexError(f"Point {p} out of bounds")
        return self.data[p.y][p.x]

    def set(self, p: Point, value: str) -> None:
        """Set character at position."""
        if self.in_bounds(p):
            self.data[p.y][p.x] = value

    def __setitem__(self, p: Point, value: str) -> None:
        """Set character at position."""
        self.set(p, value)

    def in_bounds(self, p: Point) -> bool:
        """Check if point is within grid bounds."""
        return 0 <= p.x < self.width and 0 <= p.y < self.height

    def find(self, char: str) -> Point | None:
        """Find first occurrence of character, or None if not found."""
        for y, row in enumerate(self.data):
            for x, cell in enumerate(row):
                if cell == char:
                    return Point(x, y)
        return None

    def find_all(self, char: str) -> list[Point]:
        """Find all occurrences of character."""
        result = []
        for y, row in enumerate(self.data):
            for x, cell in enumerate(row):
                if cell == char:
                    result.append(Point(x, y))
        return result

    def __iter__(self) -> Iterator[tuple[Point, str]]:
        """Iterate over all (point, character) pairs."""
        for y, row in enumerate(self.data):
            for x, cell in enumerate(row):
                yield Point(x, y), cell

    def copy(self) -> Grid:
        """Create a deep copy of the grid."""
        return Grid([row.copy() for row in self.data])

    def __str__(self) -> str:
        """Convert grid back to string."""
        return "\n".join("".join(row) for row in self.data)

    def count(self, predicate: Callable[[str], bool]) -> int:
        """Count cells matching predicate."""
        return sum(1 for _, cell in self if predicate(cell))

    def count_char(self, char: str) -> int:
        """Count occurrences of character."""
        return self.count(lambda c: c == char)

    def neighbors4_valid(self, p: Point) -> list[Point]:
        """Return valid 4-neighbors within bounds."""
        return [n for n in p.neighbors4() if self.in_bounds(n)]

    def neighbors8_valid(self, p: Point) -> list[Point]:
        """Return valid 8-neighbors within bounds."""
        return [n for n in p.neighbors8() if self.in_bounds(n)]
